### ISSUE TITLE
Support passphrase protected client certificate keys when using mTLS

### ISSUE THREAD

**Issue by cmacknz** [elastic/elastic-agent#4515]
We are adding support for mTLS between Elastic Agent and Fleet Server (or the proxy in front of Fleet Server). This requires specifying the client side private key. Currently this key must be specified on the command line which makes it easy to see. Fleet server currently supports reading the private key from a passphrase protected file. We should do the same for Elastic Agent mTLS. Support for reading these files was added in https://github.com/elastic/elastic-agent-libs/pull/109.

In the in progress mTLS implementation we do not expose command line options for reading the certificate key from a file. We need to support the `--elastic-agent-cert-key-passphrase` to support this.

- https://github.com/elastic/elastic-agent/pull/4007
- https://github.com/elastic/elastic-agent/pull/4261 

```
      --elastic-agent-cert-key-passphrase string         Path for private key passphrase file used to decrypt certificate key used in mTLS
```

Endpoint security today does not support reading keys from passphrase protected files either. Rather than make endpoint-security implement this functionality, the agent should pass the decoded private key over the control protocol (which is itself protected by mTLS on local host or local sockets/pipes). This centralizes the implementation and provides the key handling as a platform feature.

## Acceptance Critera

- Elastic Agent can be configured to use a passphrase protected private key for mTLS with a proxy or fleet server.
- Elastic Defend can connect to Fleet Server when using a passphrase protected private key for mTLS with a proxy or fleet server.

**Comment by AndersonQ** [comment]
@cmacknz @pierrehilbert @nimarezainia @jlind23 help me to clarify this issue:

> We need to support the `--fleet-server-cert-key-passphrase` to support this.

This is already implemented, the agent accepts `--fleet-server-cert-key-passphrase`. However I did not find integration tests for that or unit tests beyond checking the `--fleet-server-cert-key-passphrase` is accepted by the enroll command.

What I see missing is the client certificate key to accept a passphrase. Thus I believe what is missing on the agent side is to accept `--elastic-agent-cert-key-passphrase`.

Another thing I'd like to ask. Those configs come from the CLI as well as the policy, therefore there is also the precedence issue. I checked there are unit test on the policy change handler for `fleet.ssl.key_passphrase` and `fleet.ssl.key_passphrase_path`. 
Testing the precedence would have to be an integration test, and right now we do not have integration test with HTTPS proxy:
https://github.com/elastic/elastic-agent/blob/10d8e047258e0f3d5be1238b5e944c3105642d32/testing/integration/proxy_url_test.go#L526

Regarding Elastic Defend, shouldn't it be its own issue. Those are completely separated things, which will require different tests and more importantly can be delivered on their own.
To be honest I'm not sure how mTLS proxy/fleet-server works with Elastic Defend. Does it just uses the same client certificates as the elastic-agent?

So, in summary I believe there is at least 3 issues here:
 - add support for cli flag `--elastic-agent-cert-key-passphrase`
 - and an integration test for mTLS proxy/fleet-server, which intersects with
   - https://github.com/elastic/elastic-agent/issues/4896
   - https://github.com/elastic/elastic-agent/issues/4903
 - pass decrypted TLS client certificate to Elastic Defend through the control protocol.
   - when you say `the agent should pass the decoded private key over the control protocol`, is it requiring to extend the control protocol or the agent could just add those to the policy it passed to Elastic Defend?

**Comment by cmacknz** [comment]
> add support for cli flag --elastic-agent-cert-key-passphrase

Correct, the issue description is wrong and/or out of date.

> and an integration test for mTLS proxy/fleet-server, which intersects with

👍 Ideally we'd have a test that uses key passphrases for both the client and server side keys.

> pass decrypted TLS client certificate to Elastic Defend through the control protocol.
when you say the agent should pass the decoded private key over the control protocol, is it requiring to extend the control protocol or the agent could just add those to the policy it passed to Elastic Defend?

https://github.com/elastic/elastic-agent/blob/781f5296ae27d368ca0f5a14ae2cd34d21210a3a/internal/pkg/agent/application/fleet_server_bootstrap.go#L119-L136

Defend needs to be able to talk directly to Fleet Server, so we already inject all the fleet parameters necessary for this into their configuration. Since Defend doesn't support private key passphrases, rather than pass them the passphrase and encrypted key and require them to implement support, we can just pass them the decoded private key so they are in the same situation as regular mTLS where they are passed just the key. You don't need to change the control protocol at all.

Looking at the endpoint source code they support reading the certificate and key as long as they are in the `fleet.ssl.*` configuration block. See [here](https://github.com/elastic/endpoint-dev/blob/297f4e486547b6cc7bdc473e9f83893303cc40a6/Plugins/Policy/Lib/FleetConfig.cpp#L60) and [here](https://github.com/elastic/endpoint-dev/blob/297f4e486547b6cc7bdc473e9f83893303cc40a6/Plugins/Policy/Lib/PolicyTlsConfig.cpp#L106-L112). They are following the Filebeat ssl configuration as a spec https://www.elastic.co/guide/en/beats/filebeat/current/configuration-ssl.html#client-key. They actually parse the key passphrase but don't do anything with it.

Ideally our automated test for encrypted key files would run with Defend installed, but if that isn't possible for some reason we can get away with a test ensuring we inject the key in the expected format and a manual test.

**Comment by cmacknz** [comment]
Updated the description with the right command line options.

**Comment by AndersonQ** [comment]
I created the tickets:
 - https://github.com/elastic/elastic-agent/issues/5489
 - https://github.com/elastic/elastic-agent/issues/5490
 - https://github.com/elastic/elastic-agent/issues/5491

I believe it covers all that is here. Please have a look and see if you agree with them, particularly the out of scope tests.

**Comment by AndersonQ** [comment]
closed as it's been replaced by 

 - https://github.com/elastic/elastic-agent/issues/5489
 - https://github.com/elastic/elastic-agent/issues/5490
 - https://github.com/elastic/elastic-agent/issues/5491
